File: generator/generateMinJerkMotion.m
function builtMotion = generateMinJerkMotion(srcMotion)
    % GENERATEMINJERKMOTION  Generate a built motion for a minimum jerk primitive.
    %   builtMotion = generateMinJerkMotion(srcMotion) takes a source motion struct
    %   and produces a built motion struct with a minimum jerk trajectory from
    %   startValue to endValue.
    %
    %   Input:
    %       srcMotion - Struct with id, name, motionRef, parameters (duration,
    %                   magnitude, direction, startValue, endValue).
    %   Output:
    %       builtMotion - Struct with id, name, compositionDegree, color,
    %                     shortDisplayName, longDisplayName, magnitude, offset,
    %                     duration, flShape, frShape, rlShape, rrShape.

    % Extract parameters from source motion
    duration = srcMotion.parameters.duration;         % Seconds
    magnitude = srcMotion.parameters.magnitude;       % Newtons
    direction = srcMotion.parameters.direction;       % e.g., 'heave'
    startValue = srcMotion.parameters.startValue;     % Starting value (-1 to 1)
    endValue = srcMotion.parameters.endValue;         % Ending value (-1 to 1)

    % Constants
    Fs = 100;                                         % Sampling frequency (Hz)
    Ts = 1/Fs;                                        % Sampling period (0.01s)
    signalLen = round(duration * Fs);                 % Number of samples

    % Validate inputs
    if duration <= 0
        error('Duration must be positive');
    end
    if startValue < -1 || startValue > 1 || endValue < -1 || endValue > 1
        error('startValue and endValue must be between -1 and 1');
    end

    % Generate minimum jerk trajectory from startValue to endValue
    t = (0:Ts:duration-Ts)' / duration;              % Normalized time [0, 1]
    tau = 10*t.^3 - 15*t.^4 + 6*t.^5;                % Min jerk polynomial (0 to 1)
    y = startValue + (endValue - startValue) * tau;   % Scale from startValue to endValue
    y = y * ones(1, 4);                              % Replicate for 4 actuators

    % Apply direction to adjust actuator sequences
    flShape = y(:, 1);
    frShape = y(:, 2);
    rlShape = y(:, 3);
    rrShape = y(:, 4);
    switch lower(direction)
        case 'heave'
            % All actuators same
        case 'pitch'
            rlShape = -rlShape;                       % Rear opposite
            rrShape = -rrShape;
        case 'roll'
            frShape = -frShape;                       % Right side opposite
            rrShape = -rrShape;
        case 'fl'
            frShape = zeros(signalLen, 1);            % Only FL active
            rlShape = zeros(signalLen, 1);
            rrShape = zeros(signalLen, 1);
        case 'fr'
            flShape = zeros(signalLen, 1);            % Only FR active
            rlShape = zeros(signalLen, 1);
            rrShape = zeros(signalLen, 1);
        case 'rl'
            flShape = zeros(signalLen, 1);            % Only RL active
            frShape = zeros(signalLen, 1);
            rrShape = zeros(signalLen, 1);
        case 'rr'
            flShape = zeros(signalLen, 1);            % Only RR active
            frShape = zeros(signalLen, 1);
            rlShape = zeros(signalLen, 1);
        otherwise
            error('Invalid direction: %s', direction);
    end

    % Construct built motion struct
    builtMotion = struct();
    builtMotion.id = srcMotion.id;
    builtMotion.name = srcMotion.name;
    builtMotion.compositionDegree = 0;                % Primitive motion
    builtMotion.color = helperGenerateRandomColor();  % Random hex color
    builtMotion.shortDisplayName = 'MinJerk';
    builtMotion.longDisplayName = sprintf('Minimum Jerk Curve (%g to %g)', startValue, endValue);
    builtMotion.magnitude = magnitude;                % Overall force scale
    builtMotion.offset = 0;                           % Default offset
    builtMotion.duration = duration;
    builtMotion.flShape = flShape;                    % [signalLen x 1], startValue to endValue
    builtMotion.frShape = frShape;
    builtMotion.rlShape = rlShape;
    builtMotion.rrShape = rrShape;
end

File: generator/generate_min_jerk_motion.py
import numpy as np
from .helper_generate_random_color import helper_generate_random_color

def generate_min_jerk_motion(src_motion):
    """
    Generate a minimum-jerk trajectory from startValue to endValue.

    Parameters:
        src_motion (dict):
            - 'name'
            - 'parameters': {
                - 'duration' (float, seconds)
                - 'magnitude' (int, newtons)
                - 'direction' (str, like 'heave')
                - 'startValue' (float between -1 and 1)
                - 'endValue' (float between -1 and 1)
                }

    Returns:
        dict: builtMotion structure with actuator waveforms and metadata.
    """
    # Extract parameters
    params = src_motion['parameters']
    duration = params['duration']
    magnitude = params['magnitude']
    direction = params['direction'].lower()
    start_value = params['startValue']
    end_value = params['endValue']

    # Constants
    Fs = 100
    Ts = 1 / Fs
    signal_len = round(duration * Fs)

    # Validation
    if duration <= 0:
        raise ValueError("Duration must be positive")
    if not (-1 <= start_value <= 1 and -1 <= end_value <= 1):
        raise ValueError("startValue and endValue must be between -1 and 1")

    # Generate min-jerk trajectory (normalized time)
    t = np.arange(0, duration, Ts) / duration
    tau = 10 * t**3 - 15 * t**4 + 6 * t**5
    y = start_value + (end_value - start_value) * tau
    y = np.tile(y.reshape(-1, 1), (1, 4))  # replicate across 4 actuators

    # Apply direction shaping
    fl, fr, rl, rr = [y[:, i] for i in range(4)]
    if direction == 'heave':
        pass
    elif direction == 'pitch':
        rl *= -1
        rr *= -1
    elif direction == 'roll':
        fr *= -1
        rr *= -1
    elif direction == 'fl':
        fr[:] = rl[:] = rr[:] = 0
    elif direction == 'fr':
        fl[:] = rl[:] = rr[:] = 0
    elif direction == 'rl':
        fl[:] = fr[:] = rr[:] = 0
    elif direction == 'rr':
        fl[:] = fr[:] = rl[:] = 0
    elif direction == 'front':
        rl[:] = rr[:] = 0
    elif direction == 'rear':
        fl[:] = fr[:] = 0
    elif direction == 'left':
        fr[:] = rr[:] = 0
    elif direction == 'right':
        fl[:] = rl[:] = 0
    else:
        raise ValueError(f"Invalid direction: {direction}")

    # Construct result
    built_motion = {
        'name': src_motion['name'],
        'compositionDegree': 0,
        'color': helper_generate_random_color(),
        'shortDisplayName': 'MinJerk',
        'longDisplayName': f"Minimum Jerk Curve ({start_value:g} to {end_value:g})",
        'magnitude': magnitude,
        'offset': 0,
        'duration': duration,
        'flShape': fl.tolist(),
        'frShape': fr.tolist(),
        'rlShape': rl.tolist(),
        'rrShape': rr.tolist(),
    }

    return built_motion
